motion_tracker/utils/image_generator.py

```diff
@@ -1,137 +1,134 @@
-            to take. Dentoed as k4 in the paper, with a value of 10. 
+        crops_per_image: int
+            Number of random crops of the current frame to take.
+            Corresponds to both k3 and k4 in the paper, with a value of 10.
         batch_size: int
-            Number of items in list to be returned. 
+            Number of items in list to be returned.
+        output_width: int
+            Width in pixels of output images for start and ending image
+        output_height: int
+            Height in pixels of output images for start and ending image
+
     """
-    # To switch witch training images we draw from first in 
-    # a batch. 
-    imagenet = True
-    imagenet_gen = imagenet_generator(num_image_crops)
-    alov_gen = alov_generator(num_video_crops)
-
-    while True: 
-        
-        out_imgs = []
-        out_boxes = []
-        video_pairs = 0
-        images = 0
-        num_obs = 0
-        # Might need to add in some check or something in the 
-        # case that the num_video_crops and num_image_crops can't
-        # logically (and in an alternating way) add up to the 
-        # batch size. This while loop figures out how many pairs 
-        # of video frames to grab and how many images to grab. 
-        while num_obs < batch_size: 
-            if imagenet: 
-                images += 1
-                num_obs += num_video_crops
-            else: 
-                video_pairs += 1
-                num_obs += num_image_crops
-            imagenet = not imagenet
-
-        print(images, video_pairs)
-        
-        # Now grab the actual images and video based off the number
-        # of pairs and sets of random croppings we need. 
-        for idx in range(images): 
-            imgs, boxes = next(imagenet_gen)
-            out_imgs.append(imgs)
-            out_boxes.append(boxes)
-
-        '''
-        for idx in range(video_pairs): 
-            imgs, boxes = next(alov_gen)
-            out_imgs.append(imgs)
-            out_boxes.append(boxes)
-        '''
-
-        yield out_imgs, out_boxes
-            
-def imagenet_generator(num_image_crops):
-    """Generator yielding imagenet images and bounding boxes. 
-
-    Args: 
+
+    generators_to_draw_from = [imagenet_generator(crops_per_image)] #,
+                               # alov_gen = alov_generator(crops_per_image)]
+
+    empty_img_accumulator = np.zeros([0, output_width, output_height, 3]).astype('uint8')
+    empty_box_accumulator = np.zeros([0, 4])
+
+    while True:
+        my_output = {'start_img': empty_img_accumulator,
+                     'start_box': empty_box_accumulator,
+                     'end_img': empty_img_accumulator,
+                     'end_box': empty_box_accumulator}
+
+        # Check count of images already in batch is less than batch_size.
+        while my_output['start_img'].shape[0] < batch_size:
+            for img_source in generators_to_draw_from:
```

Comment:
This all looks great!. My only thought is on this iteration where we keep generating image croppings for the batch, potentially passed batch size, and then chop off what we don't need. I had written up the generator I wrote in a somewhat odd way, running a while loop before generating any image croppings to figure out how many times to go through the imagenet generator or the alov generator. But, this allowed us to not do the extra computation in terms of cropping more images than we need to. Maybe more importantly is that in our current case (batch size of 50), it'll end up allowing us to avoid loading an entire image that we would go through the cropping process of. I'm not sure whether this matters too much, but it's at least worth discussing. What I wrote before may or may not fit into this structure, though, so that's something to keep in mind, too. 

As another note, when we add the alov generator in, we'll need to do something to ensure that we aren't always drawing from the same generator first (unless that's something we desire). 

Really good stuff! Thanks for putting this all in the right format 👏 . I think I'm learning a good amount, and I'm loving it 😃 .